cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java

```diff
@@ -555,4 +608,37 @@ public void close() {
       baseScanner.close();
     }
   }
+
+  // scanner that matches column values based upon exact match
+  private class IndexScanner extends AbstractIndexScanner {
+    private final byte[] value;
+
+    public IndexScanner(Scanner baseScanner, byte[] column, byte[] value) {
+      super(baseScanner, column);
+      this.value = value;
+    }
+
+    @Override
+    protected boolean matches(byte[] columnValue) {
+      return Bytes.equals(columnValue, value);
+    }
+  }
+
+  // scanner that matches column values based upon range
+  private class IndexRangeScanner extends AbstractIndexScanner {
+    private final byte[] start;
+    private final byte[] end;
+
+    public IndexRangeScanner(Scanner baseScanner, byte[] column, byte[] start, byte[] end) {
+      super(baseScanner, column);
+      this.start = start;
+      this.end = end;
+    }
+
+    @Override
```

Comment:
Thinking a little more about it, there are very rare conditions where the condition is false. Only two cases I can think of is value that is a prefix of the start key, or a value of which the end key is a prefix. So I guess both checks have to stay.